aggregator/list.go

```diff
@@ -0,0 +1,343 @@
+	timer         *time.Timer
+	idBuf         bytes.Buffer
+	encoder       msgpack.AggregatedEncoder
+	toCollect     []*list.Element
+	closed        bool
+	doneCh        chan struct{}
+	wgTick        sync.WaitGroup
+	encodeFn      encodeFn
+	aggMetricFn   aggMetricFn
+	waitForFn     waitForFn
+}
+
+func newMetricList(resolution time.Duration, opts Options) *MetricList {
+	// NB(xichen): by default the flush interval is the same as metric
+	// resolution, unless the resolution is smaller than the minimum flush
+	// interval, in which case we use the min flush interval to avoid excessing
+	// CPU overhead due to flushing
+	flushInterval := resolution
+	if minFlushInterval := opts.MinFlushInterval(); flushInterval < minFlushInterval {
+		flushInterval = minFlushInterval
+	}
+
+	encoderPool := opts.BufferedEncoderPool()
+	l := &MetricList{
+		opts:          opts,
+		nowFn:         opts.ClockOptions().NowFn(),
+		log:           opts.InstrumentOptions().Logger(),
+		timeLock:      opts.TimeLock(),
+		maxFlushSize:  opts.MaxFlushSize(),
+		flushFn:       opts.FlushFn(),
+		encoderPool:   encoderPool,
+		resolution:    resolution,
+		flushInterval: flushInterval,
+		aggregations:  list.New(),
+		timer:         time.NewTimer(0),
+		encoder:       msgpack.NewAggregatedEncoder(encoderPool.Get()),
+		doneCh:        make(chan struct{}),
+	}
+	l.encodeFn = l.encoder.EncodeChunkedMetricWithPolicy
+	l.aggMetricFn = l.processAggregatedMetric
+	l.waitForFn = time.After
+
+	// Start ticking
+	if l.flushInterval > 0 {
+		l.wgTick.Add(1)
+		go l.tick()
+	}
+
+	return l
+}
+
+// Close closes the list
+func (l *MetricList) Close() {
+	l.Lock()
+	if l.closed {
+		l.Unlock()
+		return
+	}
+	l.closed = true
+	l.Unlock()
+
+	// Waiting for the ticking goroutine to finish
+	close(l.doneCh)
+	l.wgTick.Wait()
+}
+
+// PushBack adds an element to the list
+// NB(xichen): the container list doesn't provide an API to directly
+// insert a list element, therefore making it impossible to pool the
+// elements and manage their lifetimes. If this becomes an issue,
+// need to switch to a custom type-specific list implementation.
+func (l *MetricList) PushBack(value interface{}) (*list.Element, error) {
+	l.Lock()
+	if l.closed {
+		l.Unlock()
+		return nil, errListClosed
+	}
+	elem := l.aggregations.PushBack(value)
+	l.Unlock()
+	return elem, nil
+}
+
+// tick performs periodic maintenance tasks (e.g., flushing out aggregated metrics)
+func (l *MetricList) tick() {
+	defer l.wgTick.Done()
+
+	for {
+		select {
+		case <-l.doneCh:
+			return
+		default:
+			l.tickInternal()
+		}
+	}
+}
+
+func (l *MetricList) tickInternal() {
+	// NB(xichen): it is important to determine ticking start time within the time lock
+	// because this ensures all the actions before `start` have completed if those actions
+	// are protected by the same read lock
+	l.timeLock.Lock()
+	start := l.nowFn()
+	resolution := l.resolution
+	l.timeLock.Unlock()
+	alignedStart := start.Truncate(resolution)
```

Comment:
If for some reason the tick doesn't wake up in time for this to be the very next interval then we skip one it seems, is that correct?  

Just want to get a handle on the exact architecture here.